src/ucp/core/ucp_ep.c

```diff
@@ -2667,3 +2669,74 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
         *lane_map &= ~UCS_BIT(lane);
     }
 }
+
+static void ucp_ep_req_purge(ucp_ep_h ucp_ep, ucp_request_t *req,
+                             ucs_status_t status, int recursive)
+{
+    if (req->id != UCP_REQUEST_ID_INVALID) {
+        ucp_request_id_release(req);
+    }
+
+    if (req->flags & (UCP_REQUEST_FLAG_SEND_AM | UCP_REQUEST_FLAG_SEND_TAG)) {
+        ucs_assert(req->super_req == NULL);
+        ucs_assert(!(req->flags & UCP_REQUEST_FLAG_OFFLOADED));
+        ucp_request_complete_and_dereg_send(req, status);
+    } else if (req->flags &
+               (UCP_REQUEST_FLAG_RECV_AM | UCP_REQUEST_FLAG_RECV_TAG)) {
+        ucs_assert(req->super_req == NULL);
+        ucp_request_complete_and_dereg_recv_rndv(req, status);
```

Comment:
how is it possible we will try to purge a receive request? since it's not bound to an endpoint?